pulsar-broker-common/src/main/java/org/apache/pulsar/broker/authentication/metrics/AuthenticationMetrics.java

```diff
@@ -43,11 +46,36 @@ public static void authenticateSuccess(String providerName, String authMethod) {
 
     /**
      * Log authenticate failure event to the authentication metrics.
+     *
+     * This method is deprecated due to the label "reason" is a potential infinite value.
+     * @deprecated See {@link #authenticateFailure(String, String, ErrorCode)} ()}
+     * or {@link #authenticateFailure(String, String, AuthenticationException)}
+     *
      * @param providerName The short class name of the provider
      * @param authMethod Authentication method name.
      * @param reason Failure reason.
      */
+    @Deprecated
     public static void authenticateFailure(String providerName, String authMethod, String reason) {
         authFailuresMetrics.labels(providerName, authMethod, reason).inc();
     }
+
+    public static void authenticateFailure(String providerName, String authMethod,
+                                           AuthenticationException exception) {
```

Comment:
Because this is a generic metrics class meant to be useful to all authentication provider implementations, I think we should provide an easy way to provide custom "reasons". I respect that it is easy to provide a reason that is high cardinality, but I think we should expect each implementation to handle those details.

Reply:
I add an abstract class `BaseAuthenticationException`, which extends `AuthenticationException` and has an abstract method `public abstract Enum<?> getErrorCode()`, I think developers can custom their exceptions based on the `BaseAuthenticationException`, the custom exception can provide the custom error code.